src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+            {
+                lock (this) {
+                    foreach (var disposable in keep) {
+                        if (!Disposables.Contains(disposable)) {
+                            throw new InvalidOperationException("The disposable does not belong to this scope!");
+                        }
+                    }
+
+                    var toDispose = Disposables.Where(disposable => !keep.Contains(disposable)).ToList();
+                    foreach (var disposable in toDispose) {
+                        DoDispose(disposable);
+                    }
+                }
+
+                return keep;
+            }
+
+            public T DisposeEverythingBut<T>(T keep) where T : IDisposable
+            {
+                DisposeEverythingBut(new IDisposable[] { keep });
+                return keep;
+            }
+
+            public (T1 first, T2 second) DisposeEverythingBut<T1, T2>(T1 first, T2 second)
+                where T1 : IDisposable where T2 : IDisposable
+            {
+                DisposeEverythingBut(new IDisposable[] { first, second });
+                return (first, second);
+            }
+
+            public (T1 first, T2 second, T3 third) DisposeEverythingBut<T1, T2, T3>(T1 first, T2 second, T3 third)
+                where T1 : IDisposable where T2 : IDisposable where T3 : IDisposable
+            {
+                DisposeEverythingBut(new IDisposable[] { first, second, third });
+                return (first, second, third);
+            }
+
+            public void Dispose()
+            {
+                ReleaseUnmanagedResources();
+                GC.SuppressFinalize(this);
+            }
+
+            private void ReleaseUnmanagedResources()
+            {
+                lock (Singleton) {
+                    foreach (var disposable in Disposables.ToArray()) {
+                        DoDispose(disposable);
+                    }
+
+                    Singleton.RemoveDisposeScope(this);
+                }
+            }
+
+            private void DoDispose(IDisposable disposable, bool removeFromDisposables = true)
+            {
+                lock (this) {
+                    _disposeScopeManager.CurrentlyDisposing = disposable;
+                    try {
+                        if (disposable is torch.Tensor tensor) {
+                            if (!tensor.IsDisposed) {
+                                tensor.Dispose();
+                            }
+                        } else {
+                            disposable.Dispose();
+                        }
+
+                        Interlocked.Increment(ref _disposeScopeManager._disposedCount);
+                        Disposables.Remove(disposable);
+                    } finally {
+                        _disposeScopeManager.CurrentlyDisposing = null;
+                    }
+                }
+            }
+
+            ~DisposeScope()
+            {
+                ReleaseUnmanagedResources();
```

Comment:
I think it should be an error to leave a dangling instance of `DisposeScope`. Either you need it and you should eventually call `Dispose` on it, or you don't need it.

Reply:
I agree. The finalizer should simply refuse to release any resources. The only reasonable use of DisposeScopes is to put then in a using() {} construct.

Reply:
Ok, I'm confused as to whether I need a finalizer or not. I decided that I don't, so in dispose I dispose of all contained disposables and am done with it;

            public void Dispose()
            {
                foreach (var disposable in Disposables)
                {
                    DoDispose(disposable, false);
                }
                Disposables.Clear();
                _disposeScopeManager.RemoveDisposeScope(this);
            }

No suppressing anything. Makes sense?

Reply:
I think it won't hurt to have a finalizer that would throw `InvalidOperationException` unconditionally for debugging purposes.